In [110]:
import numpy as np
import pandas as pd

In [2]:
# Read SOC excel files

soc2018=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/soc_structure_2018.xlsx", header=7)
soc2018_changes = pd.read_excel("//Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/soc_2018_type_of_change_by_detail_occup.xlsx", header=8)
soc2010=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/soc_structure_2010.xls", header=11)
soc2018_dm=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/soc_2018_direct_match_title_file.xlsx",header=7)

In [3]:
# Labelling columns and cleaning up


soc2018 = soc2018.rename(columns = {'Unnamed: 4':'Occupation'})
soc2018_dm=soc2018_dm.drop(columns={'Illustrative Example'})

In [21]:
searchfor = ['C','T']
all_changes2018=soc2018_changes[soc2018_changes['Type of Change'].str.contains('|'.join(searchfor))]

In [22]:
C_changes2018=all_changes2018[~all_changes2018['Type of Change'].str.contains('T')]
T_changes2018=all_changes2018[~all_changes2018['Type of Change'].str.contains('C')]
CT_changes2018=all_changes2018[~all_changes2018['Type of Change'].str.contains('T')]

In [23]:
x=soc2018.drop(soc2018[soc2018['Detailed Occupation'].isnull()].index)

In [34]:
soc2018_dm['2018 SOC Direct Match Title']=soc2018_dm['2018 SOC Direct Match Title'].str.lower()

In [51]:
soc2018['Occupation']=soc2018['Occupation'].str.lower()

In [95]:
deleted_soc=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/soc_2010_codes_deleted_in_2018.xlsx",header=7)
deleted_soc['2010 SOC Title']=deleted_soc['2010 SOC Title'].str.lower()

In [4]:
# Reading in all the datasets from 2011-2021

df2021=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2021.xlsx",header=5)
df2020=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2020.xlsx",header=5)
df2019=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2019.xlsx",header=5)
df2018=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2018.xlsx",header=5)
df2017=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2017.xlsx",header=5)
df2016=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2016.xlsx",header=5)
df2015=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2015.xlsx",header=5)
df2014=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2014.xlsx",header=5)
df2013=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2013.xlsx",header=5)
df2012=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2012.xlsx",header=5)
df2011=pd.read_excel("/Users/terencetan/Documents/Uni stuff/Engineering Science/4YP/4YP-The-Future-of-Work/Data/Age/cpsaat11b2011.xlsx",header=5)

In [5]:
# Add in the years and rename the columns

df2021['Unnamed: 10'] = 2021
df2020['Unnamed: 10'] = 2020
df2019['Unnamed: 10'] = 2019
df2018['Unnamed: 10'] = 2018
df2017['Unnamed: 10'] = 2017
df2016['Unnamed: 10'] = 2016
df2015['Unnamed: 10'] = 2015
df2014['Unnamed: 10'] = 2014
df2013['Unnamed: 10'] = 2013
df2012['Unnamed: 10'] = 2012
df2011['Unnamed: 10'] = 2011


df=pd.concat([df2021,df2020,df2019,df2018,df2017,df2016,df2015,df2014,df2013,df2012,df2011])

df = df.rename(columns={'Unnamed: 0':'Occupation','Unnamed: 1':'Total','Unnamed: 2':'16-19','Unnamed: 3':'20-24','Unnamed: 4':'25-34','Unnamed: 5':'35-44','Unnamed: 6':'45-54','Unnamed: 7':'55-64','Unnamed: 8':'65<=','Unnamed: 9':'Median age','Unnamed: 10':'Year'})


In [6]:
# Drop redundant rows, eg. blank rows

df=df.drop(index=df.groupby('Year').head(2).index)
df=df.drop(index=df.groupby('Year').tail(2).index)
df=df[df['Occupation'].notnull()]

In [7]:
# Standardize all occupations to lowercase and remove double spaces

df['Occupation'] = df['Occupation'].str.lower()
df['Occupation']=df['Occupation'].str.replace(r'\s+', ' ', regex=True)


In [8]:
# count=df.groupby(df['Occupation'].str.lower())['Occupation'].agg('count')

In [9]:
soc2018['Occupation']=soc2018['Occupation'].str.lower()

In [70]:
joint = pd.merge(left = df[df['Year']==2018], 
         right = soc2018,
        how= 'left',
        left_on='Occupation',
        right_on='Occupation',
        indicator=True)

In [71]:
joint

Occupation    Total  16-19  \
0    management, professional, and related occupations  62436.0  436.0   
1    management, business, and financial operations...  25850.0   91.0   
2                               management occupations  18263.0   69.0   
3                                     chief executives   1573.0    0.0   
4                                     chief executives   1573.0    0.0   
..                                                 ...      ...    ...   
782          refuse and recyclable material collectors     91.0    5.0   
783                         mine shuttle car operators      1.0    0.0   
784                  tank car, truck, and ship loaders      8.0    0.0   
785                  tank car, truck, and ship loaders      8.0    0.0   
786                 material moving workers, all other     51.0    0.0   

      20-24    25-34    35-44    45-54    55-64    65<= Median age  Year  \
0    3359.0  14673.0  14572.0  13979.0  11126.0  4291.0       43.8  2018   
1    1005.0   5151.0   6069.0   6362.0   5228.0  1945.0       45.9  2018   
2     565.0   3259.0   4411.0   4666.0   3874.0  1420.0       46.6  2018   
3       3.0    147.0    349.0    467.0    413.0   193.0       51.5  2018   
4       3.0    147.0    349.0    467.0    413.0   193.0       51.5  2018   
..      ...      ...      ...      ...      ...     ...        ...   ...   
782     6.0     20.0     20.0     19.0     15.0     7.0       43.5  2018   
783     0.0      1.0      0.0      0.0      0.0     0.0          –  2018   
784     2.0      5.0      0.0      1.0      0.0     0.0          –  2018   
785     2.0      5.0      0.0      1.0      0.0     0.0          –  2018   
786     8.0      7.0     11.0     12.0     10.0     3.0       43.1  2018   

    Major Group Minor Group Broad Group Detailed Occupation     _merge  
0           NaN         NaN         NaN                 NaN  left_only  
1           NaN         NaN         NaN                 NaN  left_only  
2       11-0000         NaN         NaN                 NaN       both  
3           NaN         NaN     11-1010                 NaN       both  
4           NaN         NaN         NaN             11-1011       both  
..          ...         ...         ...                 ...        ...  
782         NaN         NaN         NaN             53-7081       both  
783         NaN         NaN         NaN                 NaN  left_only  
784         NaN         NaN     53-7120                 NaN       both  
785         NaN         NaN         NaN             53-7121       both  
786         NaN         NaN         NaN             53-7199       both  

[787 rows x 16 columns]

In [72]:
not_duplicate=joint[~joint.duplicated(['Occupation'],keep=False)]
duplicate=joint[joint.duplicated(['Occupation'],keep=False)]

In [73]:
duplicate

Occupation   Total  16-19  20-24  25-34  \
3                             chief executives  1573.0    0.0    3.0  147.0   
4                             chief executives  1573.0    0.0    3.0  147.0   
5              general and operations managers  1037.0    1.0   30.0  235.0   
6              general and operations managers  1037.0    1.0   30.0  235.0   
7                                  legislators    18.0    0.0    1.0    1.0   
..                                         ...     ...    ...    ...    ...   
775     industrial truck and tractor operators   631.0    7.0   71.0  176.0   
781  refuse and recyclable material collectors    91.0    5.0    6.0   20.0   
782  refuse and recyclable material collectors    91.0    5.0    6.0   20.0   
784          tank car, truck, and ship loaders     8.0    0.0    2.0    5.0   
785          tank car, truck, and ship loaders     8.0    0.0    2.0    5.0   

     35-44  45-54  55-64   65<= Median age  Year Major Group Minor Group  \
3    349.0  467.0  413.0  193.0       51.5  2018         NaN         NaN   
4    349.0  467.0  413.0  193.0       51.5  2018         NaN         NaN   
5    279.0  245.0  204.0   43.0       44.1  2018         NaN         NaN   
6    279.0  245.0  204.0   43.0       44.1  2018         NaN         NaN   
7      3.0    2.0    5.0    7.0          –  2018         NaN         NaN   
..     ...    ...    ...    ...        ...   ...         ...         ...   
775  127.0  137.0  101.0   13.0       39.8  2018         NaN         NaN   
781   20.0   19.0   15.0    7.0       43.5  2018         NaN         NaN   
782   20.0   19.0   15.0    7.0       43.5  2018         NaN         NaN   
784    0.0    1.0    0.0    0.0          –  2018         NaN         NaN   
785    0.0    1.0    0.0    0.0          –  2018         NaN         NaN   

    Broad Group Detailed Occupation _merge  
3       11-1010                 NaN   both  
4           NaN             11-1011   both  
5       11-1020                 NaN   both  
6           NaN             11-1021   both  
7       11-1030                 NaN   both  
..          ...                 ...    ...  
775         NaN             53-7051   both  
781     53-7080                 NaN   both  
782         NaN             53-7081   both  
784     53-7120                 NaN   both  
785         NaN             53-7121   both  

[446 rows x 16 columns]

In [61]:
duplicate[duplicate.duplicated(['Occupation'])]

Occupation   Total  16-19  20-24  25-34  \
4                             chief executives  1669.0    0.0    7.0  138.0   
6              general and operations managers  1057.0    3.0   25.0  250.0   
8                                  legislators    25.0    0.0    2.0    6.0   
10         advertising and promotions managers    56.0    1.0    1.0   22.0   
17   computer and information systems managers   744.0    0.0    7.0  128.0   
..                                         ...     ...    ...    ...    ...   
802                  transportation inspectors    35.0    1.0    1.0    9.0   
804                       passenger attendants    21.0    0.0    2.0    5.0   
807                  crane and tower operators    54.0    0.0    2.0   10.0   
810     industrial truck and tractor operators   553.0   10.0   53.0  143.0   
818  refuse and recyclable material collectors    98.0    6.0    8.0   31.0   

     35-44  45-54  55-64   65<= Median age  Year Major Group Minor Group  \
4    368.0  465.0  463.0  227.0         52  2020         NaN         NaN   
6    290.0  255.0  195.0   39.0       44.2  2020         NaN         NaN   
8      1.0    3.0    5.0    8.0          –  2020         NaN         NaN   
10    19.0   11.0    1.0    1.0       36.6  2020         NaN         NaN   
17   222.0  226.0  138.0   23.0       45.4  2020         NaN         NaN   
..     ...    ...    ...    ...        ...   ...         ...         ...   
802    6.0    8.0   10.0    1.0          –  2020         NaN         NaN   
804    2.0    3.0    4.0    4.0          –  2020         NaN         NaN   
807   12.0   17.0   10.0    1.0       45.5  2020         NaN         NaN   
810  117.0  121.0   87.0   22.0       41.5  2020         NaN         NaN   
818   21.0   17.0   12.0    4.0       38.7  2020         NaN         NaN   

    Broad Group Detailed Occupation _merge  
4           NaN             11-1011   both  
6           NaN             11-1021   both  
8           NaN             11-1031   both  
10          NaN             11-2011   both  
17          NaN             11-3021   both  
..          ...                 ...    ...  
802         NaN             53-6051   both  
804         NaN             53-6061   both  
807         NaN             53-7021   both  
810         NaN             53-7051   both  
818         NaN             53-7081   both  

[228 rows x 16 columns]

In [54]:
duplicate['Broad Group']=duplicate['Broad Group'].ffill()

<ipython-input-54-75ed9c6e470e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  duplicate['Broad Group']=duplicate['Broad Group'].ffill()


In [37]:
duplicate.groupby(duplicate['Occupation']).head(0)['Detailed Occupation'].fillna(duplicate.groupby(duplicate['Occupation']).tail(0)['Detailed Occupation'])

Series([], Name: Detailed Occupation, dtype: object)

In [30]:
not_duplicate[not_duplicate['_merge']=='both']

Occupation    Total  16-19  20-24  \
2                               management occupations  18986.0   74.0  573.0   
11                                  marketing managers    576.0    1.0   27.0   
12                                      sales managers    537.0    2.0    9.0   
13           public relations and fundraising managers    100.0    0.0    5.0   
14                    administrative services managers     60.0    0.0    1.0   
..                                                 ...      ...    ...    ...   
812  laborers and freight, stock, and material move...   2161.0  165.0  389.0   
813                     machine feeders and offbearers     22.0    2.0    9.0   
814                        packers and packagers, hand    742.0   81.0  154.0   
815                         stockers and order fillers   1652.0  205.0  357.0   
816                          pumping station operators     16.0    0.0    0.0   

      25-34   35-44   45-54   55-64    65<= Median age  Year Major Group  \
2    3413.0  4728.0  4704.0  3863.0  1630.0       46.5  2021     11-0000   
11    189.0   167.0   108.0    59.0    23.0       39.7  2021         NaN   
12     94.0   144.0   148.0   114.0    26.0       46.2  2021         NaN   
13     16.0    30.0    26.0    18.0     5.0       44.4  2021         NaN   
14      9.0    16.0    12.0    18.0     4.0       47.1  2021         NaN   
..      ...     ...     ...     ...     ...        ...   ...         ...   
812   562.0   372.0   326.0   259.0    88.0         35  2021         NaN   
813     3.0     4.0     1.0     2.0     2.0          –  2021         NaN   
814   185.0   103.0   109.0    76.0    34.0         32  2021         NaN   
815   365.0   230.0   192.0   224.0    77.0       31.9  2021         NaN   
816     3.0     5.0     5.0     2.0     2.0          –  2021         NaN   

    Minor Group Broad Group Detailed Occupation _merge  
2           NaN         NaN                 NaN   both  
11          NaN         NaN             11-2021   both  
12          NaN         NaN             11-2022   both  
13          NaN     11-2030                 NaN   both  
14          NaN         NaN             11-3012   both  
..          ...         ...                 ...    ...  
812         NaN         NaN             53-7062   both  
813         NaN         NaN             53-7063   both  
814         NaN         NaN             53-7064   both  
815         NaN         NaN             53-7065   both  
816         NaN     53-7070                 NaN   both  

[322 rows x 16 columns]

In [14]:
joint[joint['_merge']=='left_only']

Occupation    Total   16-19  \
0    management, professional, and related occupations  64744.0   420.0   
1    management, business, and financial operations...  27864.0   100.0   
101                        other financial specialists     61.0     0.0   
102               professional and related occupations  36880.0   320.0   
114             database administrators and architects    111.0     0.0   
126             other mathematical science occupations    245.0     0.0   
161                                     other drafters     93.0     0.0   
163  other engineering technologists and technician...    293.0     3.0   
177  geoscientists and hydrologists, except geograp...     42.0     0.0   
185                                other psychologists    149.0     0.0   
201  other life, physical, and social science techn...    269.0     6.0   
217     other community and social service specialists    100.0     1.0   
231       education, training, and library occupations   8946.0   120.0   
246  other educational instruction and library workers    145.0     1.0   
255                                    other designers    389.0     0.0   
294                                   other physicians    921.0     0.0   
328  nuclear medicine technologists and medical dos...     39.0     0.0   
335  dietetic technicians and ophthalmic medical te...     66.0     9.0   
344                                service occupations  24403.0  2159.0   
349                    orderlies and psychiatric aides     55.0     1.0   
360                   other healthcare support workers    164.0     3.0   
413                     janitors and building cleaners   2113.0    81.0   
419                  other grounds maintenance workers     13.0     1.0   
429  other entertainment attendants and related wor...    192.0    48.0   
430  embalmers, crematory operators, and funeral at...     14.0     0.0   
437                  other personal appearance workers     26.0     2.0   
449                       sales and office occupations  30166.0  1495.0   
572  natural resources, construction, and maintenan...  13959.0   367.0   
648                           other extraction workers     55.0     2.0   
658  electrical and electronics repairers, industri...     13.0     0.0   
677      industrial and refractory machinery mechanics    429.0     3.0   
693  other installation, maintenance, and repair wo...    182.0     5.0   
694  production, transportation, and material movin...  19309.0   825.0   
705                   other assemblers and fabricators    957.0    34.0   
717  other machine tool setters, operators, and ten...      8.0     0.0   
726            other metal workers and plastic workers    335.0     6.0   
740    other textile, apparel, and furnishings workers     14.0     1.0   
747                                  other woodworkers     22.0     1.0   
777   other production equipment operators and tenders      6.0     0.0   
778                           other production workers   1178.0    30.0   
792                  other rail transportation workers     14.0     0.0   
808    conveyor, dredge, and hoist and winch operators     35.0     1.0   
819                      other material moving workers     66.0     2.0   

      20-24    25-34    35-44    45-54    55-64    65<= Median age  Year  \
0    3267.0  15222.0  15625.0  14238.0  11394.0  4579.0       43.8  2021   
1    1052.0   5726.0   6783.0   6603.0   5411.0  2189.0       45.5  2021   
101     2.0     12.0     16.0     17.0     12.0     1.0       45.2  2021   
102  2215.0   9496.0   8842.0   7634.0   5983.0  2390.0       42.6  2021   
114     1.0     31.0     19.0     32.0     27.0     1.0       45.7  2021   
126    19.0     96.0     67.0     31.0     23.0     9.0         37  2021   
161     5.0     23.0     31.0     18.0     12.0     4.0       41.5  2021   
163    26.0     66.0     59.0     68.0     62.0    11.0       43.3  2021   
177     0.0     14.0     15.0      5.0      8.0     0.0    

In [12]:
joint[joint['_merge']=='right_only']

Occupation  Total  16-19  20-24  \
820                                      top executives    NaN    NaN    NaN   
821   advertising, marketing, promotions, public rel...    NaN    NaN    NaN   
822                        marketing and sales managers    NaN    NaN    NaN   
823                           public relations managers    NaN    NaN    NaN   
824                                fundraising managers    NaN    NaN    NaN   
...                                                 ...    ...    ...    ...   
1485                 artillery and missile crew members    NaN    NaN    NaN   
1486             command and control center specialists    NaN    NaN    NaN   
1487                                           infantry    NaN    NaN    NaN   
1488                                     special forces    NaN    NaN    NaN   
1489  military enlisted tactical operations and air/...    NaN    NaN    NaN   

      25-34  35-44  45-54  55-64  65<= Median age  Year Major Group  \
820     NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
821     NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
822     NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
823     NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
824     NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
...     ...    ...    ...    ...   ...        ...   ...         ...   
1485    NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
1486    NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
1487    NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
1488    NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   
1489    NaN    NaN    NaN    NaN   NaN        NaN   NaN         NaN   

     Minor Group Broad Group Detailed Occupation      _merge  
820      11-1000         NaN                 NaN  right_only  
821      11-2000         NaN                 NaN  right_only  
822          NaN     11-2020                 NaN  right_only  
823          NaN         NaN             11-2032  right_only  
824          NaN         NaN             11-2033  right_only  
...          ...         ...                 ...         ...  
1485         NaN         NaN             55-3014  right_only  
1486         NaN         NaN             55-3015  right_only  
1487         NaN         NaN             55-3016  right_only  
1488         NaN         NaN             55-3018  right_only  
1489         NaN         NaN             55-3019  right_only  

[670 rows x 16 columns]

2011-2019 have the same number of occupations (and presumably the same list of occupations): 572. Meanwhile 2020-2021 have 602 rows.

In [339]:
# joint[joint['_merge']=='both']

In [96]:
# df.groupby('Year').agg('count')

In [101]:
joint_deleted = pd.merge(left = df[df['Year']!=2021|2020], 
         right = deleted_soc,
        how= 'inner',
        left_on='Occupation',
        right_on='2010 SOC Title',
        indicator=True)

In [102]:
joint_deleted

Occupation  Total  16-19  20-24  25-34  \
0    public relations and fundraising managers   95.0    0.0    4.0   21.0   
1    public relations and fundraising managers   73.0    0.0    2.0   19.0   
2    public relations and fundraising managers   69.0    0.0    1.0   13.0   
3    public relations and fundraising managers   73.0    0.0    2.0   23.0   
4    public relations and fundraising managers   62.0    0.0    2.0   19.0   
..                                         ...    ...    ...    ...    ...   
496                 mine shuttle car operators    4.0    0.0    0.0    1.0   
497                 mine shuttle car operators    1.0    0.0    0.0    0.0   
498                 mine shuttle car operators    1.0    0.0    0.0    0.0   
499                 mine shuttle car operators    1.0    0.0    0.0    0.0   
500                 mine shuttle car operators    1.0    0.0    0.0    0.0   

     35-44  45-54  55-64  65<= Median age  Year 2010 SOC Group 2010 SOC Code  \
0     29.0   20.0   17.0   4.0         43  2020       Detailed       11-2031   
1     12.0   17.0   19.0   4.0       44.8  2019       Detailed       11-2031   
2     19.0   21.0   10.0   6.0         45  2018       Detailed       11-2031   
3     17.0   17.0   12.0   1.0       41.4  2017       Detailed       11-2031   
4     17.0   16.0    8.0   1.0       42.8  2016       Detailed       11-2031   
..     ...    ...    ...   ...        ...   ...            ...           ...   
496    1.0    1.0    0.0   0.0          –  2013       Detailed       53-7111   
497    1.0    0.0    0.0   0.0          –  2012          Broad       53-7110   
498    1.0    0.0    0.0   0.0          –  2012       Detailed       53-7111   
499    0.0    0.0    1.0   0.0          –  2011          Broad       53-7110   
500    0.0    0.0    1.0   0.0          –  2011       Detailed       53-7111   

                                2010 SOC Title _merge  
0    public relations and fundraising managers   both  
1    public relations and fundraising managers   both  
2    public relations and fundraising managers   both  
3    public relations and fundraising managers   both  
4    public relations and fundraising managers   both  
..                                         ...    ...  
496                 mine shuttle car operators   both  
497                 mine shuttle car operators   both  
498                 mine shuttle car operators   both  
499                 mine shuttle car operators   both  
500                 mine shuttle car operators   both  

[501 rows x 15 columns]

In [74]:
joint = pd.merge(left = df[df['Year']==2021], 
         right = df[df['Year']==2018],
        how= 'outer',
        left_on='Occupation',
        right_on='Occupation',
        indicator=True)

In [89]:
joint[joint['_merge']=='left_only'].to_excel('Compare2.xlsx', sheet_name='2021')

In [88]:
joint[joint['_merge']=='right_only'].to_excel('Compare.xlsx', sheet_name='2018')

In [353]:
df

Empty DataFrame
Columns: [Occupation, Total, 16-19, 20-24, 25-34, 35-44, 45-54, 55-64, 65<=, Median age, Year]
Index: []